kusd/validator/validator.go

```diff
@@ -259,17 +259,66 @@ func (val *Validator) restoreLastCommit() {
 func (val *Validator) init() error {
 	parent := val.chain.CurrentBlock()
 
-	val.blockNumber = parent.Number().Add(parent.Number(), big.NewInt(1))
-	val.round = 0
-
+	summary, err := val.network.VotersSummary(&bind.CallOpts{})
+	if err != nil {
+		return err
+	}
 	var start time.Time
 	if parent.NumberU64() == 0 {
 		start = time.Now()
+
+		// first validator set
+		count, err := val.network.GetVoterCount(&bind.CallOpts{})
+		if err != nil {
+			return err
+		}
+		validators := make([]*types.Validator, count.Uint64())
+		for i := int64(0); i < count.Int64(); i++ {
+			validator, err := val.network.GetVoterAtIndex(&bind.CallOpts{}, big.NewInt(i))
+			if err != nil {
+				return err
+			}
+			validators[i] = types.NewValidator(validator.Addr, validator.Deposit.Uint64(), big.NewInt(0))
+		}
+		val.validators = types.NewValidatorSet(validators)
+		val.validatorsSummary = summary
 	} else {
 		start = time.Unix(parent.Time().Int64(), 0)
+
+		// new validator set
+		if val.validatorsSummary != summary {
+			count, err := val.network.GetVoterCount(&bind.CallOpts{})
+			if err != nil {
+				return err
+			}
+			val.validatorsSummary = summary
+			validators := make([]*types.Validator, count.Uint64())
+			for i := int64(0); i < count.Int64(); i++ {
+				validator, err := val.network.GetVoterAtIndex(&bind.CallOpts{}, big.NewInt(i))
+				if err != nil {
+					return err
+				}
+				var weight *big.Int
+				if val.validators.Contains(validator.Addr) {
+					// old validator
+					old := val.validators.Get(validator.Addr)
+					weight = old.Weight()
+				} else {
+					// new validator
+					weight = big.NewInt(0)
+				}
+				validators[i] = types.NewValidator(validator.Addr, validator.Deposit.Uint64(), weight)
+			}
+			val.validators = types.NewValidatorSet(validators)
+			val.validatorsSummary = summary
```

Comment:
This whole code (from the `GetVoterCount...`) it's exactly the same as the previous code except for the weight. I think this version can be used in both places because it will never find an `old` validator, so it'll use `0` weight in all cases, right? If so, this could be refactored to something like `val.buildValidators()`.